File: 786-SearchInASortedArrayOfUnknownSize/786-SearchInASortedArrayOfUnknownSize.py
# Last updated: 2/9/2026, 9:54:43 PM
# """
# This is ArrayReader's API interface.
# You should not implement it, or speculate about its implementation
# """
#class ArrayReader:
#    def get(self, index: int) -> int:

class Solution:
    def search(self, reader: 'ArrayReader', target: int) -> int:
        
        # set boundaries
        
        left = 0
        right = 1
        
        while target > reader.get(right):
            right = right * 2
        
        
        # perform binary search        
        while left <= right:
            
            mid = ((right - left) // 2) + left
            
            if reader.get(mid) == target:
                return mid
            elif reader.get(mid) < target:
                left = mid + 1
            else:
                right = mid - 1
        
        return -1
         

File: 786-SearchInASortedArrayOfUnknownSize/786-SearchInASortedArrayOfUnknownSize.java
// Last updated: 2/9/2026, 9:54:47 PM
/**
 * // This is ArrayReader's API interface.
 * // You should not implement it, or speculate about its implementation
 * interface ArrayReader {
 *     public int get(int index) {}
 * }
 */

class Solution {
    public int search(ArrayReader reader, int target) {

        int i = 0;
        
        while (reader.get(i) < Math.pow(2, 31) - 1) {
            if (reader.get(i) == target) return i;
            i++;
        }

        return -1;
    }
}